File: solutions/java/hamming/1/src/main/java/Hamming.java
import java.util.Iterator;

public class Hamming {

    private String leftStrand, rightStrand;

    public Hamming(String leftStrand, String rightStrand) {
        if (leftStrand.length() == 0 && rightStrand.length() > 0) {
            throw new IllegalArgumentException("left strand must not be empty.");
        }
        if (leftStrand.length() > 0 && rightStrand.length() == 0) {
            throw new IllegalArgumentException("right strand must not be empty.");
        }
        if (leftStrand.length() != rightStrand.length()) {
            throw new IllegalArgumentException("leftStrand and rightStrand must be of equal length.");
        }
        this.leftStrand = leftStrand;
        this.rightStrand = rightStrand;
    }

    public int getHammingDistance() {
        int difference = 0;
        for (int i = 0; i < leftStrand.length(); i++) {
            if (leftStrand.charAt(i) != rightStrand.charAt(i)) {
                difference++;
            }
        }
        return difference;
    }
}

File: solutions/ocaml/hamming/1/hamming.ml
open Base

type nucleotide = A | C | G | T

let rec distance dna1 dna2 count =
  match (dna1, dna2) with
    | (x::xs, y::ys) -> distance xs ys (if x != y then count + 1 else count) 
    | _ -> count
    

let hamming_distance dna1 dna2 =
  match (List.is_empty dna1, List.is_empty dna2) with 
  | (true, false) -> Error "left strand must not be empty"
  | (false, true) -> Error "right strand must not be empty"
  | _ -> 
      if List.length dna1 <> List.length dna2 then 
        Error "left and right strands must be of equal length"
      else Ok (distance dna1 dna2 0)